### PR TITLE
executor: XSelectIndexExec fetch index and get data concurrently

### PR DESCRIPTION
**Pull request by tiancaiamao**
Previously, we must wait for `fetchHandles` to fetch all index data, before scanning any table data.

Now it's changed to a pipelined fashion, `go fetchHandles()` would run in background and put task into `e.tasks`,  so fetch index and get table data can run concurrently.

### CODE REVIEW

File: executor/new_executor_xapi.go [hunk 1]
```diff
@@ -229,69 +226,79 @@ func (e *NewXSelectIndexExec) indexRowToTableRow(handle int64, indexRow []types.
...
-			concurrency = int(limitCount/int64(BaseLookupTableTaskSize) + 1)
-		}
-		log.Debugf("[TIME_INDEX_TABLE_CONCURRENT_SCAN] start %d workers", concurrency)
-		e.runTableTasks(concurrency)
+		idxResult.Fetch()
+
+		// use a background goroutine to fetch index, put the result in e.tasks.
+		// e.tasks serves as a pipeline, so fetch index and get table data would
+		// run concurrency.
+		e.tasks = make(chan *lookupTableTask, 10)
```
Comment (hanfei1991) [inline executor/new_executor_xapi.go]: "10" should be changed to "default concurrency", should it ?

Comment (tiancaiamao) [inline executor/new_executor_xapi.go]: in the thread pool model, there is a queue,  and a set of threads which are the workers.
mostly when we meet the word "concurrency", it means the number of workers.
while the "10" here is the capacity of queue...

File: executor/executor_xapi.go [hunk 2]
```diff
@@ -193,23 +204,21 @@ func convertIndexRangeTypes(ran *plan.IndexRange, fieldTypes []*types.FieldType)
 	return nil
 }
 
+// extractHandlesFromIndexResult gets some handles from SelectResult.
+// it should be called in a loop until nil is returned.
```
Comment (hanfei1991) [inline executor/executor_xapi.go]: it -> It

File: executor/new_executor_xapi.go [hunk 3]
```diff
@@ -229,69 +226,79 @@ func (e *NewXSelectIndexExec) indexRowToTableRow(handle int64, indexRow []types.
...
+	defer close(workCh)
+	limitCount := -1
+	if e.indexPlan.LimitCount != nil {
+		limitCount = int(*e.indexPlan.LimitCount)
 	}
-	idxResult.Fetch()
-	handles, err := extractHandlesFromIndexResult(idxResult)
-	if err != nil {
-		return nil, errors.Trace(err)
+	concurrency := 2
```
Comment (hanfei1991) [inline executor/new_executor_xapi.go]: parameterize it

File: executor/new_executor_xapi.go [hunk 4]
```diff
@@ -229,69 +226,79 @@ func (e *NewXSelectIndexExec) indexRowToTableRow(handle int64, indexRow []types.
...
+				e.indexOrder[h] = i
+			}
+		}
+
+		tasks := e.buildTableTasks(handles)
+		if limitCount == -1 {
+			limitCount = len(tasks)
+		}
+		for concurrency < limitCount {
+			go e.pickAndExecTask(workCh)
```
Comment (hanfei1991) [inline executor/new_executor_xapi.go]: Will two goroutines pick same look-up-table-task?

**Comment by hanfei1991** [comment]
Any bench result ?

File: executor/new_executor_xapi.go (hunk 4 above)
Comment (tiancaiamao) [inline executor/new_executor_xapi.go]: of course not.
`workCh` is a channel, one data in it (`*lookupTableTask`) can't be dequeue twice.

Comment (hanfei1991) [inline executor/new_executor_xapi.go]: Many goroutines fetches task from one task. Will some goroutines fetch many tasks but some goroutines fetch few tasks?

File: executor/new_executor_xapi.go [hunk 5]
```diff
@@ -229,69 +226,79 @@ func (e *NewXSelectIndexExec) indexRowToTableRow(handle int64, indexRow []types.
...
+			for i, h := range handles {
+				e.indexOrder[h] = i
+			}
+		}
+
+		tasks := e.buildTableTasks(handles)
+		if limitCount == -1 {
+			limitCount = len(tasks)
+		}
+		for concurrency < limitCount {
```
Comment (hanfei1991) [inline executor/new_executor_xapi.go]: The concurrency can't be decided by limitCount.

File: executor/new_executor_xapi.go [hunk 6]
```diff
@@ -229,69 +226,72 @@ func (e *NewXSelectIndexExec) indexRowToTableRow(handle int64, indexRow []types.
...
-	if err != nil {
-		return nil, errors.Trace(err)
+func (e *NewXSelectIndexExec) fetchHandles(idxResult xapi.SelectResult, ch chan<- *lookupTableTask) {
+	defer close(ch)
+
+	workCh := make(chan *lookupTableTask)
+	defer close(workCh)
+	concurrency := 2
+	for i := 0; i < concurrency; i++ {
+		go e.pickAndExecTask(workCh)
```
Comment (coocood) [inline executor/new_executor_xapi.go]: Why create two goroutines before build table tasks?

File: executor/new_executor_xapi.go [hunk 7]
```diff
@@ -229,69 +226,72 @@ func (e *NewXSelectIndexExec) indexRowToTableRow(handle int64, indexRow []types.
...
+	for {
+		handles, err := extractHandlesFromIndexResult(idxResult)
+		if err != nil || handles == nil {
+			e.tasksErr = errors.Trace(err)
+			return
+		}
+
+		if !e.indexPlan.OutOfOrder {
+			// Save the index order.
+			e.indexOrder = make(map[int64]int)
```
Comment (coocood) [inline executor/new_executor_xapi.go]: This branch is not covered by tests.
`indexOrder` should belong to a partial result of index request, but it is a single field in the executor and used by all `pickAndExecTask` workers,

File: executor/new_executor_xapi.go (hunk 6 above)
Comment (tiancaiamao) [inline executor/new_executor_xapi.go]: those two goroutines serve as background workers, they will wait for tasks from `workCh`.
... or should we create them until some task put to `workCh` ?

Comment (coocood) [inline executor/new_executor_xapi.go]: We can create goroutines when tasks have been built, the number of tasks can be one.

**Comment by tiancaiamao** [comment]
PTAL @shenli

File: executor/new_executor_xapi.go [hunk 8]
```diff
@@ -257,25 +260,39 @@ func (e *NewXSelectIndexExec) nextForDoubleRead() (*Row, error) {
 func (e *NewXSelectIndexExec) fetchHandles(idxResult xapi.SelectResult, ch chan<- *lookupTableTask) {
 	defer close(ch)
 
-	workCh := make(chan *lookupTableTask, 10)
+	workCh := make(chan *lookupTableTask, 1)
```
Comment (shenli) [inline executor/new_executor_xapi.go]: Why make the channel with only one buffer?

Comment (tiancaiamao) [inline executor/new_executor_xapi.go]: We don't need many buffer. 

Every time sending to channel may block, a new goroutine is created.
At line 290:

```
		select {
			case workCh <- task:
			default:
				go e.pickAndExecTask(workCh)
				concurrency++
				workCh <- task
		}
```

I'm considering not reuse worker and `go goroutine()` directly for each `lookupTableTask`, but I'm not sure.

File: executor/new_executor_xapi.go [hunk 9]
```diff
@@ -228,70 +223,79 @@ func (e *NewXSelectIndexExec) indexRowToTableRow(handle int64, indexRow []types.
...
+				totalHandles,
+				concurrency)
+			return
+		}
+
+		totalHandles += len(handles)
+		tasks := e.buildTableTasks(handles)
+		for _, task := range tasks {
+			if concurrency < len(tasks) {
+				go e.pickAndExecTask(workCh)
```
Comment (shenli) [inline executor/new_executor_xapi.go]: We should set a limitation for the concurrency.

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@coocood @hanfei1991 @zimulala PTAL

File: executor/new_executor_xapi.go [hunk 10]
```diff
@@ -338,59 +352,36 @@ func (e *NewXSelectIndexExec) buildTableTasks(handles []int64) {
...
-	if e.indexPlan.Desc && !e.supportDesc {
-		// Reverse tasks sizes.
-		for i := 0; i < len(taskSizes)/2; i++ {
-			j := len(taskSizes) - i - 1
-			taskSizes[i], taskSizes[j] = taskSizes[j], taskSizes[i]
+
+	var indexOrder map[int64]int
+	if !e.indexPlan.OutOfOrder {
+		// Save the index order.
+		indexOrder = make(map[int64]int)
```
Comment (ngaut) [inline executor/new_executor_xapi.go]: Make with capacity

**Comment by tiancaiamao** [comment]
@ngaut PTAL

File: executor/new_executor_xapi.go [hunk 11]
```diff
@@ -228,70 +223,89 @@ func (e *NewXSelectIndexExec) indexRowToTableRow(handle int64, indexRow []types.
...
+	defer close(workCh)
+
+	var concurrency int
+	addWorker(e, workCh, &concurrency)
+
+	totalHandles := 0
+	startTs := time.Now()
+	for {
+		handles, err := extractHandlesFromIndexResult(idxResult)
+		if err != nil || handles == nil {
```
Comment (coocood) [inline executor/new_executor_xapi.go]: Is it possible that a partial result returns nil handles, but there still other handles in other partial result to be extracted?

**Comment by tiancaiamao** [comment]
PTAL @coocood

File: executor/executor_xapi.go [hunk 12]
```diff
@@ -206,20 +206,24 @@ func convertIndexRangeTypes(ran *plan.IndexRange, fieldTypes []*types.FieldType)
 }
 
 // extractHandlesFromIndexResult gets some handles from SelectResult.
-// It should be called in a loop until nil is returned.
-func extractHandlesFromIndexResult(idxResult xapi.SelectResult) ([]int64, error) {
-	subResult, err := idxResult.Next()
-	if err != nil {
-		return nil, errors.Trace(err)
+// It should be called in a loop until finish or error.
+func extractHandlesFromIndexResult(idxResult xapi.SelectResult) (handles []int64, finish bool, err error) {
```
Comment (coocood) [inline executor/executor_xapi.go]: finish -> finished

File: executor/executor_xapi.go [hunk 13]
```diff
@@ -206,20 +206,24 @@ func convertIndexRangeTypes(ran *plan.IndexRange, fieldTypes []*types.FieldType)
...
 	}
 	if subResult == nil {
-		return nil, nil
+		finish = true
+		return
 	}
-	subHandles, err := extractHandlesFromIndexSubResult(subResult)
-	if err != nil {
-		return nil, errors.Trace(err)
+	subHandles, e1 := extractHandlesFromIndexSubResult(subResult)
```
Comment (coocood) [inline executor/executor_xapi.go]: `handles, e1 := extractHandlesFromIndexSubResult(subResult)` will do.

File: executor/new_executor_xapi.go [hunk 14]
```diff
@@ -280,8 +280,8 @@ func (e *NewXSelectIndexExec) fetchHandles(idxResult xapi.SelectResult, ch chan<
 	totalHandles := 0
 	startTs := time.Now()
 	for {
-		handles, err := extractHandlesFromIndexResult(idxResult)
-		if err != nil || handles == nil {
+		handles, finish, err := extractHandlesFromIndexResult(idxResult)
+		if err != nil || finish == true {
```
Comment (coocood) [inline executor/new_executor_xapi.go]: `if err != nil || finished {` will do.

**Comment by coocood** [comment]
LGTM

**Pull request merged**